src/QsCompiler/CompilerOptimizations/CallableInlining.fs

```diff
@@ -0,0 +1,183 @@
+/// If no such specialization is found, returns None.
+let private tryGetImpl callable (functors: Functors) =
+    callable.Specializations
+    |> Seq.tryFind (fun s -> s.Kind = functors.toSpecKind)
+    |> Option.map (fun s -> s.Implementation)
+
+/// Tries to find a provided implementation for the given callable with the given specialization kind.
+/// Returns None if unable to find a provided implementation of the desired kind.
+let private tryGetProvidedImpl callable functors =
+    match tryGetImpl callable functors, tryGetImpl callable functors.withAdjoint with
+    | Some (Provided (specArgs, body)), _
+    | Some (Generated SelfInverse), Some (Provided (specArgs, body)) -> Some (specArgs, body)
+    | _ -> None
+
+
+/// Stores all the data needed to inline a callable
+type private InliningInfo = {
+    functors: Functors
+    callable: QsCallable
+    arg: TypedExpression
+    specArgs: QsArgumentTuple
+    body: QsScope
+}
+
+/// Tries to construct an InliningInfo from the given expression.
+/// Returns None if the expression is not a callable invocation that can be inlining.
+let private tryGetInliningInfo callables expr =
+    maybe {
+        let! functors, callable, arg = trySplitCall callables expr
+        let! specArgs, body = tryGetProvidedImpl callable functors
+        return { functors = functors; callable = callable; arg = arg; specArgs = specArgs; body = body }
+    }
+
+
+/// Recursively finds all the callables that could be inlined into the given scope.
+/// Includes callables that could be indirectly called or inlined into this scope.
+/// Mutates the given HashSet by adding all the found callables to the set.
+/// Is used to prevent inlining recursive functions into themselves forever.
+let rec private findAllCalls (callables: Callables) (scope: QsScope) (found: HashSet<QsQualifiedName>): unit =
+    scope |> findAllBaseStatements |> Seq.iter (function
+        | QsExpressionStatement ex ->
+            match tryGetInliningInfo callables ex.Expression with
+            | Some ii ->
+                // Only recurse if we haven't processed this callable yet
+                if found.Add ii.callable.FullName then
+                    findAllCalls callables ii.body found
+            | None -> ()
+        | _ -> ())
+
+/// Returns whether the given callable could eventually inline the given callable.
+/// Is used to prevent inlining recursive functions into themselves forever.
+let private cannotReachCallable (callables: Callables) (scope: QsScope) (cannotReach: QsQualifiedName) =
+    let mySet = HashSet()
+    findAllCalls callables scope mySet
+    not (mySet.Contains cannotReach)
+
+    
+/// The SyntaxTreeTransformation used to inline callables
+type internal CallableInliner(callables) =
+    inherit OptimizingTransformation()
+
+    // The current callable we're in the process of transforming
+    let mutable currentCallable: QsCallable option = None
+    // The list of new statements to add
```

Comment:
This code is currently unfinished, and is unneeded if we only care about inlining callables that are invoked as ExpressionStatements. If we do care about inlining callables that are used as part of other expressions, then I will have to finish implementing this code. Otherwise, I can remove this code. How much do we care about inlining non-expression-statements?